DRA plugin: handle gRPC serving failures

bart0sh opened the pull request:
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

It implements a way to catch fatal errors that happen in go-routines. DRA plugin can catch them
and perform graceful shutdown. The implementation demonstrates this approach on catching
grpc.Server.Serve() error, which was not processed before, resulting in a DRA driver running,
but absolutely non functional after grpc.Server.Serve failure.

#### Special notes for your reviewer:

If this PR is accepted I'm going to modify example driver to fail on serving failures.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

bart0sh committed:
e2e_node: Refactor DRA tests to use variadic options

Refactor the DRA e2e_node test helpers and test cases to accept
variadic kubeletplugin.Option arguments.

This change improves test flexibility and maintainability, allowing
new options to be passed in the future without requiring widespread
code changes.

There are no functional changes to test coverage or behavior.

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-linter-hints
/test pull-kubernetes-unit-windows-master
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/cc @nojnhuh

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -274,15 +274,15 @@ func PluginSocket(name string) Option {
 }
 
-// PluginListener configures how to create the registrar socket.
+// PluginListener configures how to create the socket listener.
```
The listener for which socket?

"plugin" and "endpoint" are both ambiguous. Let's be clear about which socket we are dealing with.

I also like "create socket" better than "create socket listener". The main purpose of the function is to create a socket; the returned `net.Listener` is then just an implementation detail.

The Go documentation is similar: "net.Listen" is described as "announces on the local network address", not as "creates a listener".

So I would fix the copy-and-paste error with:
```
PluginListener configures how to create the DRA service socket
```

pohly replied:
> For either plugin(service) or registrar.

Really? There's a separate `RegistrarListener` for the registrar socket.

> This could be understood as it's for creating plugin(service) socket only, which is not correct.

It is correct. The parameter of `PluginListener` is only used for one socket.

bart0sh commented:
/cc @SergeyKanzhelev @lauralorenz

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -430,6 +430,18 @@ func DRAService(enabled bool) Option {
 
+// ErrorChannel allows the plugin to send errors to the caller. This is
+// useful to report errors that happen in goroutines. The caller is expected
+// to read from the channel and handle errors appropriately.
```
Can you add guidance for DRA drivers for what "handle errors appropriately" means?

Log them and continue? Log and exit because all errors are fatal? Are there special errors that are merely informative (probably not, but worth clarifying)?

bart0sh replied:
The only usage of this is handling grpc.Server.Serve errors. Below explanation suggests performing a graceful shutdown. In other cases (feature cases) it could be different.

pohly replied:
It says "log *this* error and perform a graceful shutdown" (emphasis mine) - but it doesn't say how the caller can detect *this* error.

Either we need to document how it can detect that error, or recommend that it does that action for *all* errors.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -430,6 +430,18 @@ func DRAService(enabled bool) Option {
+// to read from the channel and handle errors appropriately.
+// For example, this channel can be used to report errors returned by the
+// grpcserver.Serve. Plugins then can use this error to perform a graceful shutdown.
+func ErrorChannel(errorChannel chan error) Option {
```
```suggestion
func ErrorChannel(errorChannel chan<- error) Option {
```

The helper is only meant to write to this channel, not read from it.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/nonblockinggrpcserver.go:
```diff
@@ -78,7 +78,10 @@ func startGRPCServer(logger klog.Logger, grpcVerbosity int, unaryInterceptors []
 		err := s.server.Serve(listener)
 		if err != nil {
-			logger.Error(err, "GRPC server failed")
+			logger.Error(err, "gRPC server failed to serve")
```
Let's log here only if we don't pass on the error. Otherwise it might get logged twice.

pohly reviewed (approved):
/approve

My comments on the API have been addressed. As @nojnhuh was tagged earlier and this is related to some discussions that your were having, I'll defer to him for the LGTM.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -430,6 +430,18 @@ func DRAService(enabled bool) Option {
+// plugin cannot return an error directly.
+// DRA plugins are expected to read from the error channel, log received errors,
+// and perform cleanup and graceful shutdown operations.
+func ErrorChannel(errorChannel chan<- error) Option {
```
The implementation of `ErrorHandler` can call `os.Exit`.

nojnhuh replied:
Yeah I think binding this to the context would make the most sense from a driver author perspective. Ideally something close to:
1. Start plugin
2. Wait for context done (either from OS signal or some fatal background error)
3. Shutdown

bart0sh committed:
e2e_node: DRA: add errorOnCloseListener

Introduce a mock net.Listener for tests that triggers a controlled
error on Close, enabling reliable simulation of gRPC server failures
in test scenarios.

aojea commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -489,7 +489,7 @@ type Helper struct {
 // If the plugin will be used to publish resources, [NodeName]
 // is also mandatory.
-func Start(ctx context.Context, plugin DRAPlugin, opts ...Option) (result *Helper, finalErr error) {
+func Start(ctx context.Context, cancelMainContext context.CancelCauseFunc, plugin DRAPlugin, opts ...Option) (result *Helper, finalErr error) {
```
having a context and then plumbing a cancel function that later is merged with the context is really confusing

bart0sh replied:
I'd be happy to provide a context only, but I don't see any way to get cancel function(that is a result of context.WithCancel* on the upper level) from it.

bart0sh replied:
Would not merging it make it better? I thought that one cancel function is better than two. Or we're talking about 2 different things here?

pohly replied:
That said "from a driver author perspective" - my interpretation was that the driver author then will do that in their code.

pohly replied:
`context.CancelCauseFunc` = `func(error)`. So what this does is adding a second callback that has the exact same signature as the existing `ErrorHandler` interface method, it just gets called for different errors and is provided through some different means (parameter instead of interface implementation).  I don't think we should be doing this.

It's also completely undocumented, which is a problem not only for DRA driver developers, but also for reviewers.

aojea commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/nonblockinggrpcserver.go:
```diff
@@ -78,13 +78,17 @@ func startGRPCServer(logger klog.Logger, grpcVerbosity int, unaryInterceptors []
 		err := s.server.Serve(listener)
 		if err != nil {
-			logger.Error(err, "GRPC server failed")
+			if cancelMainContext != nil {
+				cancelMainContext(fmt.Errorf("gRPC server failed to serve on %s: %w", endpoint.path(), err))
+			} else {
+				logger.Error(err, "gRPC server failed to serve", "endpoint", endpoint.path())
+			}
```
why a grpc server error should cancel a context that is up in the chain of context cancellation?

pohly commented:
/retitle DRA plugin: handle gRPC serving failures

pohly committed:
DRA: handle grpc.Server.Serve failures

Added an errHandler callback to startGRPCServer for handling errors.
Replaced direct logging of server failures with calls to the
provided error handler. This change ensures that errors are surfaced
and handled consistently.

Co-authored-by: Ed Bartosh <eduard.bartosh@intel.com>

bart0sh committed:
e2e_node: DRA: test handling fatal serving failures

Added an e2e_node test to verify that the DRA plugin and
registration services cancel provided context when handling
fatal gRPC serving errors.

pohly commented on test/e2e_node/dra_test.go:
```diff
@@ -457,6 +459,48 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
+				listenerOptionFun(getListener),
+				kubeletplugin.ErrorHandler(func(ctx context.Context, err error) {
+					cancel(err)
+				}),
```
We can't put them into a separate package because that package would need access to the unexported options struct. We would need to move that struct to `kubeletplugin/internal` and export it there. This is too much work just to inform DRA driver developers that an option is merely for testing purposes. A comment should suffice for that.

But I am still not convinced that this option is needed. The upper layers could accept `opts ...any` and then accept a mixture of their own options and `kubeletplugin.Option`. When processing the options, each layer can react to its own options and pass down only the rest. It's not compile-time-safe, but for testing purposes that should be fine.

pohly replied:
> Moving testing options to testing.go would be enough in my opinion.

Readers of the `go doc k8s.io/dynamic-resource-allocation/kubeletplugin` don't see where an option is defined.

It's not just the extra option, but also the additional code in the package that doesn't belong into DRA drivers.

bart0sh committed:
e2e: DRA: support test and public options

Refactor StartPlugin and related test helpers to accept a variadic
list of options of any type, allowing both public and test-specific
options to be passed.

bart0sh committed:
e2e: DRA: enable test plugin to cancel main context on error

Add a cancelMainContext field to ExamplePlugin and options. This
function allows the plugin to cancel an upper-level context when a
background error occurs. The HandleError method now calls this
function if it is set.

This enables DRA drivers to perform a graceful shutdown when a fatal
background error triggers context cancellation.

bart0sh commented on test/e2e_node/dra_test.go:
```diff
@@ -830,31 +839,19 @@ func newRegistrar(ctx context.Context, clientSet kubernetes.Interface, nodeName,
-		)
-		datadir = dir
-	}
+	datadir := path.Join(kubeletplugin.KubeletPluginsDir, driverName)
```
PluginDataDirectoryPath and PluginSocket options are used only in the testing code. Would it make sense to convert them into testing options and obtain datadir from them if they're set?

bart0sh replied:
Wouldn't `datadir` parameter conflict with above mentioned options? As we don't have access to the options data we can't verify them on the test level.

bart0sh replied:
I'd leave creation of the data dir to the driver. I don't think this is a right place for it.

pohly replied:
> Wouldn't datadir parameter conflict with above mentioned options? 

The caller has to ensure that it passes a datadir parameter which is consistent with the options.

>  I'd leave creation of the data dir to the driver.

Normally the driver shouldn't need to create them. That's part of the driver deployment (hostpath + mount), which corresponds to this code here.

bart0sh replied:
done

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -115,15 +115,22 @@ type DRAPlugin interface {
+	// achieved by canceling the main context of the DRA driver.
+	//
+	// Fatal errors can be distinguished from recoverable errors via
+	//    errors.Is(err, kubeletplugin.ErrRecoverable)
```
@bart0sh: thoughts on this?

bart0sh replied:
I renamed it.

pohly reviewed (changes requested):
Almost there...

bart0sh commented on test/e2e_node/dra_test.go:
```diff
@@ -829,31 +885,33 @@ func newRegistrar(ctx context.Context, clientSet kubernetes.Interface, nodeName,
+//   - clientSet: Kubernetes client interface for interacting with the cluster.
+//   - nodeName: The name of the node where the plugin will run.
+//   - driverName: The name of the DRA driver.
+//   - datadir: The directory for the DRA socket; if empty, a default path is used.
```
done

pohly reviewed (approved):
/lgtm
/approve
/hold

In case that @nojnhuh has any further comments about using this in the example driver.

nojnhuh commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -119,20 +119,43 @@ type DRAPlugin interface {
+// and should cause the process to exit. Use with:
+//
+//	errors.Is(err, ErrRecoverable)
+var ErrRecoverable = errors.New("recoverable error")
```
Should we invert this to be `ErrIrrecoverable` if it positively indicates a fatal error?

nojnhuh replied:
Oh I think I misread this.

nojnhuh reviewed (approved):
LGTM. I integrated this with the example driver and it's working the way I expect.

/hold cancel
